Create database via UI when ActiveRecord::NoDatabaseError

hahmed opened the pull request:
### Summary

Create the database via the UI when database has not been created in development mode. 

Related: https://discuss.rubyonrails.org/t/pending-migration-error-usability-concerns/75330/3

I have split out the errors for connection and no database errors and added the button to create the database.

<img width="1344" alt="Screenshot 2020-07-17 at 21 09 08" src="https://user-images.githubusercontent.com/658005/87827135-21f8c580-c872-11ea-9b10-ff21744fd93e.png">

<img width="1346" alt="Screenshot 2020-07-17 at 21 09 59" src="https://user-images.githubusercontent.com/658005/87827148-26bd7980-c872-11ea-9ee7-04b19d338630.png">

jonathanhefner commented:
This looks like a job for [`ActionableError`](https://github.com/rails/rails/blob/0f6a4f5895a4366abe6951ec7fad99cec031cee9/activesupport/lib/active_support/actionable_error.rb)!  :superhero:  (See [`PendingMigrationError`](https://github.com/rails/rails/blob/0f6a4f5895a4366abe6951ec7fad99cec031cee9/activerecord/lib/active_record/migration.rb#L133-L138) for an example of usage.)

hahmed commented:
@jonathanhefner I have uploaded a screenshot of how the 2 new interfaces look. 

Would love some feedback on how I can make the pages friendly and welcoming?

jonathanhefner commented:
This is only my opinion, but I'm not sure I see the benefit of dedicated error pages.  I feel like a combination of the following would accomplish our goals:

* automatically handling errors, e.g. automatically running migrations in dev / test
* using `ActionableError`
* improving error messages on the exceptions themselves
* improving the friendliness of the shared error template, if necessary

(I think it would be good to break these into separate PRs, too.)

hahmed commented:
Agreed 👍

I'm going to update this PR and description shortly, to only include the action for creating the database via the interface, similar to migrations.

tgxworld commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -185,6 +185,21 @@ class PreparedStatementInvalid < ActiveRecordError
+    include ActiveSupport::ActionableError
+
+    action "Create database" do
+      ActiveRecord::Tasks::DatabaseTasks.create_current
```
This would only create the default database. The `NoDatabaseError` could be thrown for a secondary database.

hahmed replied:
Sounds good, will make that update.

hahmed replied:
@tgxworld I tested this out and it seems to be working for me, here is the configuration I used:

```
development:
  primary:
    database: my_primary_database
    adapter: postgresql
  primary_replica:
    database: my_primary_database
    replica: true
    adapter: postgresql
  animals:
    database: my_animals_database
    migrations_paths: db/animals_migrate
    adapter: postgresql
  animals_replica:
    database: my_animals_database
    replica: true
    adapter: postgresql

# test db
test:
  <<: *default
  database: askaway_test
```

Here is the output from the console, showing I created 3 databases:
```
Started POST "/rails/actions" for 127.0.0.1 at 2020-07-09 16:50:11 +0100
   (80.8ms)  CREATE DATABASE "my_primary_database" ENCODING = 'utf8'
  ↳ /Users/haroon/projects/rails/activerecord/lib/active_record/log_subscriber.rb:109:in `debug'
Created database 'my_primary_database'
   (78.7ms)  CREATE DATABASE "my_animals_database" ENCODING = 'utf8'
  ↳ /Users/haroon/projects/rails/activerecord/lib/active_record/log_subscriber.rb:109:in `debug'
Created database 'my_animals_database'
   (80.5ms)  CREATE DATABASE "askaway_test" ENCODING = 'unicode'
  ↳ /Users/haroon/projects/rails/activerecord/lib/active_record/log_subscriber.rb:109:in `debug'
Created database 'askaway_test'
```

hahmed replied:
If you can let me know what configs you used to repro, I can try and fix... 🙂

tgxworld replied:
Sorry you're right. I just realized that when the error is thrown, the active connection_pool will be that of the missing database.

rafaelfranca reviewed (commented):
I don't think we can tell people to run this command every time this specific exception happens.

The database may not exist because they changed by mistake the `config/database.yml` to point to a different database. So to resolve the issue they should not be running `bin/rails db:create` instead they should be fixing the database name.

I also not sure if this error would not also happen if the username or the password is wrong, or the database host is also wrong. If the same exception happens for those cases, we have more reason to not add this button by default, neither show an error message telling people to create the db since that would not fix the exception, it would probably just give another error when trying to create the database with the wrong credentials.

I feel we need to understand better the cases where this error happen.

hahmed commented:
I had a look into this and agree with you. I think this PR would be better suited to adding a more detailed error message.

> The database may not exist because they changed by mistake the config/database.yml to point to a different database. 

Perhaps a more detailed error like this would be user friendly?

```
We could not find your database your_app_database_name_here, which can be found in the database configuration file config/database.yml.

What you could do to resolve this issue:

- Has your database name inadvertently been changed? If so, your config file would need to be the name of your database.
- Perhaps you have never created a database for this app, or deleted it? If so you could run `bin/rails db:create` to create your database.
```

> I also not sure if this error would not also happen if the username or the password is wrong, or the database host is also wrong.

I have tested this scenario on Postgres, when the username is wrong we get the error:

```
PG::ConnectionBad
#<PG::ConnectionBad: FATAL:  role "randomuserrole" does not exist>
```

When the hostname is wrong, we get the same exception as above:

```
#<PG::ConnectionBad: could not connect to server: Operation timed out
Is the server running on host "notlocalhost" (92.242.132.24) and accepting
TCP/IP connections on port 5432?
```

We could capture the ConnectionBad error and provide more specific feedback if that sounds feasible?

```
Looks like you have an error with your hostname or username/password. What you could do to resolve this issue:

- Please check the username and password match your database settings. 
- Please ensure your hostname is correct.
```

For sqlite3 when the username/password or hostname was random, it still worked.

hahmed commented on activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:
```diff
@@ -8,7 +8,10 @@
 module ActiveRecord
   module ConnectionHandling # :nodoc:
-    ER_BAD_DB_ERROR = 1049
+    ER_BAD_DB_ERROR        = 1049
```
Source: https://github.com/brianmario/mysql2/blob/ca08712c6c8ea672df658bb25b931fea22555f27/lib/mysql2/error.rb#L16

hahmed commented:
@rafaelfranca wonder if you have any more feedback on this, I ended up looking at your suggestion about better understanding the error message and I have posted the results in my previous comment -- https://github.com/rails/rails/pull/39723#issuecomment-658690335.

With mysql we get back a number of different error messages, which I have highlighted here https://github.com/rails/rails/pull/39723/files#r456110152 and only show the button to create the database via the UI when we get a bad connection error. 

Similarly to postgresql we only show the create database button in the UI when the database name exists in the error message from pg here https://github.com/rails/rails/pull/39723/files#diff-cc31fc1dd1e78db64638200f710b2f59R40

rafaelfranca commented:
Let's move this forward. Can you rebase?

hahmed committed:
Add a more detailed error messages for connection issues to pg and
mysql. Allow database to be create via a button on the UI for some
connection issues.

hahmed commented:
I rebased and found an error on CI (below), which seems to be related to the stuff I done, however it's also failing when I pull down master and run the postgresql tests. Will have a look...

```
Failure:
ActiveRecord::ConnectionAdapters::PostgreSQLAdapterTest#test_reconnection_error [/Users/haroon/projects/rails/activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb:56]:
ActiveRecord::ConnectionNotEstablished expected but nothing was raised.


rails test Users/haroon/projects/rails/activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb:25
```

hahmed commented:
I have rebased and the CI errors seem unrelated...

hahmed commented:
@rafaelfranca I ended up reverting the changes for the formatted error message, as it broke a few other tests. I think I will add an attribute onto the exception's to opt-into the simple_formatted display, that way other tests are not impacted. Will follow that up in another PR if that's ok?

Let me know if there is there anything else I need to do on this PR.

hahmed commented:
@rafaelfranca is there anything else I can do on this PR to push it forward? 😀

I split out the changes for the formatting which got merged, this PR now only deals with creating the database.

zzak commented:
@hahmed This change seems like we were close to merging it before, so going to re-open and give it some thought.

zzak commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -202,6 +202,52 @@ class PreparedStatementInvalid < ActiveRecordError
+
+        To resolve this issue:
+
+        - Not created a database for this app, or deleted it? You may need to create your database.
```
```suggestion
        - Did you create the database for this app, or delete it? You may need to create your database.
```

zzak commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -202,6 +202,52 @@ class PreparedStatementInvalid < ActiveRecordError
+        To resolve this issue:
+
+        - Not created a database for this app, or deleted it? You may need to create your database.
+        - Database name changed? Check your database.yml config has the correct database name.
```
```suggestion
        - Has the database name changed? Check your database.yml config has the correct database name.
```

zzak commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -202,6 +202,52 @@ class PreparedStatementInvalid < ActiveRecordError
+    class << self
+      def hostname_error(hostname)
+        DatabaseConnectionError.new(<<~MSG)
+        We are having an issue connecting with your hostname: #{hostname}.\n
```
```suggestion
        There is an issue connecting with your hostname: #{hostname}.\n
```

zzak commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -202,6 +202,52 @@ class PreparedStatementInvalid < ActiveRecordError
+      def hostname_error(hostname)
+        DatabaseConnectionError.new(<<~MSG)
+        We are having an issue connecting with your hostname: #{hostname}.\n
+        Please check your database configuration or ensure there is a connection open to your hostname.
```
```suggestion
        Please check your database configuration and ensure there is a valid connection to your database.
```

zzak commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -202,6 +202,52 @@ class PreparedStatementInvalid < ActiveRecordError
+
+      def username_error(username)
+        DatabaseConnectionError.new(<<~MSG)
+        We are having an issue connecting to your database with your username/password, username: #{username}.\n
```
```suggestion
        There is an issue connecting to your database with your username/password, username: #{username}.\n
```

zzak commented:
@hahmed Is the button still there to run the rake task for you? If not can you update the screenshots (after updating the knits I suggested)? :bow:

hahmed committed:
Apply suggestions from code review

Co-authored-by: Zachary Scott <zzakscott@gmail.com>

hahmed commented:
Thanks for reviewing @zzak ♥️

Updated screenshot has the "Create database" button:

<img width="1440" alt="Screenshot 2021-05-17 at 20 21 13" src="https://user-images.githubusercontent.com/658005/118544675-8b28be00-b74d-11eb-96ad-537a6edcb8e5.png">

zzak commented:
@hahmed Thanks! I will check this out and give it a closer look soon 🙏
